distribution/lib/Standard/Base/0.0.0-dev/src/Any.enso

```diff
@@ -108,26 +108,23 @@ type Any
     == self that =
         # If there is No_Such_Conversion, then `self` and `that` are probably
         # host or polyglot values, so we just compare them with the default comparator.
-        eq_self = Panic.catch No_Such_Conversion (Comparable.from self) _-> Default_Unordered_Comparator
-        eq_that = Panic.catch No_Such_Conversion (Comparable.from that) _-> Default_Unordered_Comparator
-        case Meta.is_same_object eq_self Incomparable of
-            True -> False
-            False ->
-                similar_type = Meta.is_same_object eq_self eq_that
-                case similar_type of
-                    False -> False
+        eq_self = Panic.catch No_Such_Conversion (Comparable.from self) _-> Default_Comparator
+        eq_that = Panic.catch No_Such_Conversion (Comparable.from that) _-> Default_Comparator
+        similar_type = Meta.is_same_object eq_self eq_that
+        case similar_type of
+            False -> False
+            True ->
+                case Meta.is_same_object eq_self Default_Comparator of
                     True ->
-                        case eq_self.is_ordered of
-                            True ->
-                                res = eq_self.compare self that
-                                case res of
-                                    Ordering.Equal -> True
-                                    _ -> False
-                            False ->
-                                res = eq_self.equals self that
-                                case res of
-                                    Nothing -> False
-                                    _ -> res
+                        # Comparable.equals_builtin can return Nothing when self and that
+                        # are incomparable.
+                        case Comparable.equals_builtin self that of
+                            True -> True
+                            _ -> False
```

Comment:
minor nitpick but I'm a bit surprised we do the `case-of` check instead of just delegating to `Default_Comparator.compare` which could be the place that will then call `Comparable.equals_builtin self that`?

Reply:
`Default_Comparator.compare x y` has to handle three cases - `x==y`, `x<y` or `x>y` - which is implemented with three nested case expressions. If `x` and `y` are neither, i.e., are _not equal_, all of these three cases have to be evaluated. So this is, once again, because of the performance - it is just a shortcut.

Reply:
OK.

I'm really surprised we have these 3 builtins. IMO one builtin `compare x y` returning -1, 0, +1 or `null` could work just as well or maybe even better.

Like, if you are comparing two Text values that have 1M characters of length (an exaggerated example but well...), assume the difference is at second to last character. The equals check will have to iterate the whole string and find the diff at the end, to determine they are not equal. But are they greater or smaller? Now we have to iterate the whole string once again...

Instead with `compare` approach we'd do this at one time.
I think even for integers/floats the JVM should be able to boil down a == + < into a single three-way compare instruction (although [citation needed]).

Reply:
We don't have 3 builtins for comparisons, we have only 2 - `EqualsNode` (`Comparable.equals_builtin`) and `LessThanNode` (`Comparable.less_than_builtin`). Merging their functionality into one is inappropriate because many types that can be compared with `==`,  cannot be compared with `<` - Just compare the number of specializations we have in `EqualsNode` and compare it to the number of specializations in `LessThanNode`. Moreover, `LessThanNode` now has a different return value, as it can return `Nothing` for values that are incomparable.

The example with a string with 1M characters is, indeed, exaggerated. Besides, merging `EqualsNode` and `LessThanNode` in one single node might bring us some performance improvement. But comparing it to other low hanging fruits, like problems with nested case expressions, it would be negligible.